include/seqan3/alignment/matrix/detail/aligned_sequence_builder.hpp

```diff
@@ -27,94 +30,49 @@
+ * with seqan3::gapped alphabet over the underlying alphabet of the range will be used.
  */
-template <std::ranges::viewable_range fst_rng_t,
-          std::ranges::viewable_range sec_rng_t,
-          typename traits_t = aligned_sequence_builder_default_traits<fst_rng_t, sec_rng_t>>
+template <std::ranges::viewable_range fst_rng_t, std::ranges::viewable_range sec_rng_t>
 class aligned_sequence_builder
 {
 private:
-
-    static_assert(std::constructible_from<typename traits_t::fst_aligned_t,
-                                     decltype(std::declval<fst_rng_t>() | views::slice(0, 1))>,
-                  "The given sequence type cannot be transformed to a aligned_sequence. It needs to be a views::slice "
-                  "over the original type.");
-
-    static_assert(std::constructible_from<typename traits_t::sec_aligned_t,
-                                     decltype(std::declval<sec_rng_t>() | views::slice(0, 1))>,
-                  "The given sequence type cannot be transformed to a aligned_sequence. It needs to be a views::slice "
-                  "over the original type.");
-
-    using fst_aligned_t = typename traits_t::fst_aligned_t; //!< The aligned sequence type for the first range.
-    using sec_aligned_t = typename traits_t::sec_aligned_t; //!< The aligned sequence type for the second range.
+    //!\brief The aligned sequence type for the first range.
+    using fst_aligned_t =
+        lazy_conditional_t<
+            is_instantiable_with_v<gap_decorator, decltype(std::declval<fst_rng_t>() | views::slice(0, 1))>,
+            lazy<gap_decorator, decltype(std::declval<fst_rng_t>() | views::slice(0, 1))>,
+            lazy<std::vector, gapped<std::ranges::range_value_t<fst_rng_t>>>>;
+    //!\brief The aligned sequence type for the second range.
+    using sec_aligned_t =
+        lazy_conditional_t<
+            is_instantiable_with_v<gap_decorator, decltype(std::declval<sec_rng_t>() | views::slice(0, 1))>,
+            lazy<gap_decorator, decltype(std::declval<sec_rng_t>() | views::slice(0, 1))>,
+            lazy<std::vector, gapped<std::ranges::range_value_t<sec_rng_t>>>>;
+
+    static_assert(seqan3::aligned_sequence<fst_aligned_t>,
+                  "fst_aligned_t is required to model seqan3::aligned_sequence!");
+    static_assert(seqan3::aligned_sequence<sec_aligned_t>,
+                  "sec_aligned_t is required to model seqan3::aligned_sequence!");
 
 public:
 
     //!\brief The result type when building the aligned sequences.
     struct result_type
     {
-        matrix_coordinate begin; //!< The coordinate where the trimmed first and second sequence begin.
-        matrix_coordinate end; //!< The coordinate where the trimmed first and second sequence end.
-        std::pair<fst_aligned_t, sec_aligned_t> alignment; //!< The alignment over the trimmed sequences.
+        matrix_coordinate begin{}; //!< The coordinate where the trimmed first and second sequence begin.
```

Comment:
I did changed the behavior and the naming. Also I wasn't happy with begin and end since it might indicate some itertor or something like that. I hope the renaming reflects the logic of this more.